### PR TITLE
Groundwork for hold action changes

### PR DESCRIPTION
**Pull request by sethalves**
- adjusted locks around various bullet action related code
- added a terse form of server-side entity-edit logging and associated domain-server setting control
- moved some methods from MyAvatar to Avatar
- fixed some problems that might have kept actions from being deleted

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7ef5fc9613cf190cb03482b8dad8391b675535ef available at: https://gist.github.com/f65ba6433bba426bee1d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c96f7562f277b0459abf61a11d7205cb3926c65a available at: https://gist.github.com/f65ba6433bba426bee1d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d833fb08d74f365323822f5e0dd8553c42c70b90 available at: https://gist.github.com/f65ba6433bba426bee1d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8195145dc787b952d3230f82f9b154b30b42d968 available at: https://gist.github.com/f65ba6433bba426bee1d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e571cb7ca44384121fc48e7868a411198ec770a3 available at: https://gist.github.com/f65ba6433bba426bee1d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a78728c96b0f447f9a13e74df3ae6101c8096d2e available at: https://gist.github.com/f65ba6433bba426bee1d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 64c5cde2d2aee33acc43a8487401cc6dc44d9204 available at: https://gist.github.com/f65ba6433bba426bee1d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c884f466bc3d7977abc977f09522d4a3d91689b7 available at: https://gist.github.com/f65ba6433bba426bee1d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6083.dmg

File: interface/src/avatar/AvatarManager.cpp [hunk 1]
```diff
@@ -343,3 +343,11 @@ void AvatarManager::updateAvatarRenderStatus(bool shouldRenderAvatars) {
...
     }
 }
+
+
+AvatarSharedPointer AvatarManager::getAvatarBySessionID(const QUuid& sessionID) {
+    if (sessionID == _myAvatar->getSessionUUID()) {
+        return std::static_pointer_cast<Avatar>(_myAvatar);
+    }
+    return getAvatarHash()[sessionID];
+}
```
Comment (AndrewMeadows) [inline interface/src/avatar/AvatarManager.cpp]: I don't see where AvatarManager::getAvatarBySessionID() is ever used.  Why was it added?

Comment (sethalves) [inline interface/src/avatar/AvatarManager.cpp]: It's used the in actual grab changes, which I haven't included in this PR.

File: libraries/entities/src/EntityItemProperties.cpp [hunk 2]
```diff
@@ -1519,3 +1519,254 @@ void EntityItemProperties::setSimulationOwner(const QByteArray& data) {
...
+        out += "zPNeighborID";
+    }
+
+    getAnimation().listChangedProperties(out);
+    getAtmosphere().listChangedProperties(out);
+    getSkybox().listChangedProperties(out);
+    getStage().listChangedProperties(out);
+
+    return out;
+}
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityItemProperties.cpp]: This looks like even more boiler plate for adding entity properties but it is just a **debug** method and it isn't technically necessary to add new properties to it.  Fine.

File: libraries/physics/src/ObjectAction.cpp [hunk 3]
```diff
@@ -24,20 +24,33 @@ ObjectAction::~ObjectAction() {
...
+            QUuid myID;
+            withWriteLock([&]{
+                ownerEntity = _ownerEntity.lock();
+                _active = false;
+                myID = getID();
+            });
             if (ownerEntity) {
-                ownerEntity->removeAction(nullptr, getID());
+                ownerEntity->removeAction(nullptr, myID);
             }
```
Comment (AndrewMeadows) [inline libraries/physics/src/ObjectAction.cpp]: The computation of **myID** here is protected by a lock but it is not inside a lock in other contexts (such as **ObjectActionOffset::getArguments()**).  So I'm wondering... is the lock required when calling **getID()** or not?

File: examples/controllers/handControllerGrab.js [hunk 4]
```diff
@@ -441,7 +444,9 @@ function MyController(hand, triggerAction) {
         this.currentObjectTime = now;
 
         // this doubles hand rotation
-        var handChange = Quat.multiply(Quat.slerp(this.handPreviousRotation, handRotation, DISTANCE_HOLDING_ROTATION_EXAGGERATION_FACTOR), Quat.inverse(this.handPreviousRotation));
+        var handChange = Quat.multiply(Quat.slerp(this.handPreviousRotation, handRotation,
+                                                  DISTANCE_HOLDING_ROTATION_EXAGGERATION_FACTOR),
+                                       Quat.inverse(this.handPreviousRotation));
```
Comment (AndrewMeadows) [inline examples/controllers/handControllerGrab.js]: Wow, you can slerp() with a coefficient larger than 1.0.  You learn something new every day.

**Pull request merged**